### ISSUE TITLE
[TODOs] ElasticDL on SQLFlow

### ISSUE THREAD

**Issue by terrytangyuan** [sql-machine-learning/sqlflow#664]
Initial setup and skeleton code has been added in https://github.com/sql-machine-learning/sqlflow/pull/663. Below is a list of additional work items for ElasticDL integration with SQLFlow:

# Training
- [x] Parse `SELECT ... FROM ...`, generate code to read from ODPS table and write it to [RecordIO](https://github.com/wangkuiyi/recordio) files.
- [x] Execute the above generated data conversion script before submitting `elasticdl` commands. This depends on https://github.com/wangkuiyi/elasticdl/issues/1039.
- [x] Parse `COLUMN ... LABEL ...` to figure out `<input_shape>` and `<num_classes>`.
- [x] Allow `custom_model()` to take `<input_shape>` and `<num_classes>` that will be passed to `elasticdl train` via `--params` later.
- [x] Pass additional parameters from `WITH` clause to `custom_model()`'s instantiation, such as `optimizer` and `loss`.
- [x] Construct `dataset_fn`: parse `COLUMN` clause and then pass column names, shapes, and types for features and labels to `dataset_fn`'s feature description that will be used in `tf.io.parse_single_example()`. For this MVP, column names can be obtained from `SELECT ... LABEL ...`. Each feature columns will be of shape `[1]` and of type `tf.float32` while label column is of shape `[1]` and of type `tf.int64` for classification problems and `tf.float32` for regression problems.
- [x] Skip support for feature transformation functions such as `NUMERIC` or `BUCKET` in `COLUMN` clause for now. Add some checks/warnings to handle this in codegen.
- [x] Parse `INTO` clause and pass the information to `--outputs` argument in ElasticDL high-level API.
- [x] Parse run-time configurations that can be passed to `elasticdl train` command later.
- [x] Pass generated model definition file name and other args to `elasticdl train` command.

# Prediction
- [x] Generate `PredictionOutputsProcessor` that writes prediction result to ODPS.
- [x] Parse `USING` clause to get the name to the trained model to be used to make predictions.
- [x] Pass the generated args and model definition file to `elastic predict` command.

# Others
- [x] Register allowed ElasticDL-specific keywords in `WITH` clause.
- [x] Register ElasticDL filler.
- [x] Add integration test, including the following steps: 
a. SQLFLlow generates model definition files and high-level `elasticdl train` command
b. `elasticdl train` submits a job to `minikube` on Travis and the job reads/writes ODPS data (references to the current SQLFlow test setup for ODPS [test_maxcompute.sh](https://github.com/sql-machine-learning/sqlflow/blob/develop/scripts/test_maxcompute.sh) and [main_test.go](https://github.com/sql-machine-learning/sqlflow/blob/5e7adc1ececc92995423097328eb3ca63c27f99b/cmd/sqlflowserver/main_test.go#L319-L356))
c. Check whether job is successful

**Comment by tonyyang-svail** [comment]
Hi @terrytangyuan, thank you for submitting this solid TODO list. Could you also describe the integration test?

**Comment by terrytangyuan** [comment]
@tonyyang-svail Sure, just added that to the list.

**Comment by typhoonzero** [comment]
About the `WITH` attributes, I'd like to first define some overall rules in the parser, like attribute names should all follow: 

- `model.[attr]` define attributes passed to the model initialization phase.
- `train.[attr]` define attributes when calling `estimator.train` ( or maybe `trainspec` ) or `model.fit` for keras.
- `eval.[attr]` attributes for evaluation.
- `engine.[attr]` for distributed submission attributes like `num_workers`.

All submitters should follow these rules and attributes will first go through a "validator", to keep the user experience the same.

**Comment by terrytangyuan** [comment]
ElasticDL support has been added.
